app/scripts/controllers/permissions/loggerMiddleware.js

```diff
@@ -0,0 +1,169 @@
+      request: cloneObj(request),
+      requestTime: Date.now(),
+      response: null,
+      responseTime: null,
+      success: null,
+    }
+    commitActivity(activityEntry)
+    return activityEntry
+  }
+
+  function addResponse (activityEntry, response, time) {
+    if (!response) {
+      return
+    }
+    activityEntry.response = cloneObj(response)
+    activityEntry.responseTime = time
+    activityEntry.success = !response.error
+  }
+
+  function commitActivity (entry) {
+    const logs = store.getState()[logStoreKey]
+    if (logs.length > LOG_LIMIT - 2) {
+      logs.pop()
+    }
+    logs.push(entry)
+    store.updateState({ [logStoreKey]: logs })
+  }
+
+  function getRequestedMethods (request) {
+    if (
+      !request.params ||
+      typeof request.params[0] !== 'object' ||
+      Array.isArray(request.params[0])
+    ) {
+      return null
+    }
+    return Object.keys(request.params[0])
+  }
+
+  function logHistory (requestedMethods, origin, result, time, isEthRequestAccounts) {
+    let accounts, entries
+    if (isEthRequestAccounts) {
+      accounts = result
+      const accountToTimeMap = accounts.reduce((acc, account) => ({ ...acc, [account]: time }), {})
+      entries = { 'eth_accounts': { accounts: accountToTimeMap, lastApproved: time } }
+    } else {
+      entries = result
+        ? result
+          .map(perm => {
+            if (perm.parentCapability === 'eth_accounts') {
+              accounts = getAccountsFromPermission(perm)
+            }
+            return perm.parentCapability
+          })
+          .reduce((acc, m) => {
+            if (requestedMethods.includes(m)) {
+              if (m === 'eth_accounts') {
+                const accountToTimeMap = accounts.reduce((acc, account) => ({ ...acc, [account]: time }), {})
+                acc[m] = { lastApproved: time, accounts: accountToTimeMap }
+              } else {
+                acc[m] = { lastApproved: time }
+              }
+            }
+            return acc
+          }, {})
+        : {}
+    }
+
+    if (Object.keys(entries).length > 0) {
+      commitHistory(origin, entries)
+    }
+  }
+
+  function commitHistory (origin, entries) {
+    const history = store.getState()[historyStoreKey] || {}
+    const newOriginHistory = {
+      ...history[origin],
+      ...entries,
+    }
+
+    if (history[origin] && history[origin]['eth_accounts'] && entries['eth_accounts']) {
+      newOriginHistory['eth_accounts'] = {
+        lastApproved: entries['eth_accounts'].lastApproved,
+        accounts: {
+          ...history[origin]['eth_accounts'].accounts,
+          ...entries['eth_accounts'].accounts,
+        },
+      }
+    }
+
+    history[origin] = newOriginHistory
+
+    store.updateState({ [historyStoreKey]: history })
+  }
+}
+
+// the call to clone is set to disallow circular references
+// we attempt cloning at a depth of 3 and 2, then return a
+// shallow copy of the object
+function cloneObj (obj) {
```

Comment:
I'm not sure I follow why this is something we want to do. Could we use `fast-deep-clone` instead?

Out of curiosity: do request objects typically contain circular references? What might those references be?

Reply:
Paraphrasing from @rekmarks here, but it turns out that it's a defensive measure to prevent maliciously formatted requests from having a performance impact upon MetaMask.

`fast-deepclone` won't work because it isn't browser compatible, but `rfdc` with the `circular` option enabled should do the trick. I'm not opposed to leaving this in as-is though - we can adjust this up some other time.